ui/src/main.rs

```diff
@@ -104,362 +100,92 @@ pub static WINDOW_CMD_CH: Lazy<WindowManagerCmdChannels> = Lazy::new(|| {
-        div {
-            id: "app-wrap",
-            div {
-                class: "titlebar disable-select",
-                id: if cfg!(target_os = "macos") {""}  else {"lockscreen-controls"},
-                onmousedown: move |_| { desktop.drag(); },
-                Topbar_Controls {},
-            },
-            match *page.current() {
-                AuthPages::Unlock => rsx!(UnlockLayout { page: page.clone(), pin: pin.clone() }),
-                AuthPages::CreateAccount => rsx!(CreateAccountLayout { page: page.clone(), pin: pin.clone() }),
-                _ => panic!("invalid page")
-            }
-        }
-    ))
+    #[route("/files")]
+    FilesLayout {},
 }
 
-fn get_extensions() -> Result<HashMap<String, UplinkExtension>, Box<dyn std::error::Error>> {
-    fs::create_dir_all(&STATIC_ARGS.extensions_path)?;
-    let mut extensions = HashMap::new();
-
-    let mut add_to_extensions = |dir: fs::ReadDir| -> Result<(), Box<dyn std::error::Error>> {
-        for entry in dir {
-            let path = entry?.path();
-            log::debug!("Found extension: {:?}", path);
-
-            match UplinkExtension::new(path.clone()) {
-                Ok(ext) => {
-                    if ext.cargo_version() != extensions::CARGO_VERSION
-                        || ext.rustc_version() != extensions::RUSTC_VERSION
-                    {
-                        log::warn!("failed to load extension: {:?} due to rustc/cargo version mismatch. cargo version: {}, rustc version: {}", &path, ext.cargo_version(), ext.rustc_version());
-                        continue;
-                    }
-                    log::debug!("Loaded extension: {:?}", &path);
-                    extensions.insert(ext.details().meta.name.into(), ext);
-                }
-                Err(e) => {
-                    log::error!("Error loading extension: {:?}", e);
-                }
-            }
-        }
+fn app(cx: Scope) -> Element {
+    // 1. Make sure the warp engine is turned on before doing anything
+    bootstrap::use_warp_runner(cx);
 
-        Ok(())
+    // 2. Guard the app with the auth
+    let auth = use_state(cx, || AuthPages::Unlock);
+    let AuthPages::Success(identity) = auth.get() else {
+        return render! { AuthGuard { page: auth.clone() }};
     };
 
-    let user_extension_dir = fs::read_dir(&STATIC_ARGS.extensions_path)?;
-    add_to_extensions(user_extension_dir)?;
+    // 3. Make sure global context is setup before rendering anything downstream
+    bootstrap::use_boostrap(cx, identity)?;
 
-    if STATIC_ARGS.production_mode {
-        let uplink_extenions_path = common::get_extensions_dir()?;
-        let uplink_extensions_dir = fs::read_dir(uplink_extenions_path)?;
-        add_to_extensions(uplink_extensions_dir)?;
+    // 4. Throw up a loading screen until our assets are ready
+    if use_loaded_assets(cx).value().is_none() {
+        return render! { LoadingWash {} };
```

Comment:
Similar to above, we do loading states and such in a one-way flow to ensure downstream components are okay to render.